File: apps/node-web-push/src/services/push.service.js
const webPush = require('web-push');
const { config } = require('../config.js');
const { subscriptionService } = require('./subscription.service.js');

// Configure Web Push with VAPID details
webPush.setVapidDetails(
  config.vapidSubject,
  config.vapidKeys.publicKey,
  config.vapidKeys.privateKey
);

const pushService = {
  /**
   * Dispatch push notification payload to all registered subscriptions.
   * @param {import('../types.js').PushNotificationPayload} payloadData
   * @returns {Promise<{ successCount: number, failCount: number }>}
   */
  async sendPushToAll(payloadData) {
    const payload = JSON.stringify(payloadData);
    let successCount = 0;
    let failCount = 0;

    const subscriptions = subscriptionService.getAllSubscriptions();

    for (const [id, sub] of subscriptions.entries()) {
      try {
        await webPush.sendNotification(sub, payload);
        successCount++;
      } catch (err) {
        const error = /** @type {{ message?: string, statusCode?: number }} */ (err);
        console.error(`[Server] Failed to send push to ID ${id}:`, error?.message || err);
        failCount++;

        // Remove stale/expired subscriptions (410 Gone / 404 Not Found)
        if (error?.statusCode === 410 || error?.statusCode === 404) {
          subscriptionService.removeSubscription(id);
        }
      }
    }

    console.log(`[Server] Push dispatch complete. Success: ${successCount}, Failed: ${failCount}`);
    return { successCount, failCount };
  },
};

module.exports = { pushService };

File: apps/golang-web-push/services/push_service.go
package services

import (
	"encoding/json"
	"log"
	"net/http"
	"sync"

	webpush "github.com/SherClockHolmes/webpush-go"

	"github.com/FarrelAD/WaraWara-Web/apps/golang-web-push/config"
	"github.com/FarrelAD/WaraWara-Web/apps/golang-web-push/models"
)

// PushService dispatches notifications to push services using webpush-go.
type PushService struct{}

var NotificationService = &PushService{}

// SendPushToAll sends the given payload to all currently registered subscriptions concurrently.
func (p *PushService) SendPushToAll(payload models.PushNotificationPayload) models.DispatchResult {
	subs := SubService.GetAll()
	result := models.DispatchResult{
		Total: len(subs),
	}

	if len(subs) == 0 {
		return result
	}

	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		log.Printf("[PushService ERROR] Failed to serialize notification payload: %v\n", err)
		result.Failed = len(subs)
		return result
	}

	cfg := config.AppConfig

	var wg sync.WaitGroup
	var mu sync.Mutex

	for _, sub := range subs {
		wg.Add(1)
		go func(s models.PushSubscription) {
			defer wg.Done()

			webpushSub := &webpush.Subscription{
				Endpoint: s.Endpoint,
				Keys: webpush.Keys{
					P256dh: s.Keys.P256dh,
					Auth:   s.Keys.Auth,
				},
			}

			resp, sendErr := webpush.SendNotification(payloadBytes, webpushSub, &webpush.Options{
				Subscriber:      cfg.VapidSubject,
				VAPIDPublicKey:  cfg.VapidPublicKey,
				VAPIDPrivateKey: cfg.VapidPrivateKey,
				TTL:             30,
			})

			mu.Lock()
			defer mu.Unlock()

			if sendErr != nil {
				log.Printf("[PushService ERROR] Delivery error for %s: %v\n", s.Endpoint, sendErr)
				result.Failed++
				return
			}
			defer resp.Body.Close()

			if resp.StatusCode == http.StatusCreated || resp.StatusCode == http.StatusOK {
				result.Successful++
			} else if resp.StatusCode == http.StatusGone || resp.StatusCode == http.StatusNotFound {
				// Subscription expired or uninstalled -> prune from storage
				log.Printf("[PushService INFO] Subscription expired (status %d). Pruning endpoint %s\n", resp.StatusCode, s.Endpoint)
				SubService.RemoveByEndpoint(s.Endpoint)
				result.Pruned++
			} else {
				log.Printf("[PushService WARNING] Unexpected push service status %d for endpoint %s\n", resp.StatusCode, s.Endpoint)
				result.Failed++
			}
		}(sub)
	}

	wg.Wait()
	return result
}
